packages/contracts/test/utils/test_with_reference.ts

```diff
@@ -64,39 +102,17 @@ export async function testWithReferenceFuncAsync(
     testFuncAsync: (...args: any[]) => Promise<any>,
     values: any[],
 ): Promise<void> {
-    let expectedResult: any;
-    let expectedErr: string | undefined;
-    try {
-        expectedResult = await referenceFuncAsync(...values);
-    } catch (e) {
-        expectedErr = e.message;
-    }
-    let actualResult: any | undefined;
-    try {
-        actualResult = await testFuncAsync(...values);
-        if (!_.isUndefined(expectedErr)) {
-            throw new Error(
-                `Expected error containing ${expectedErr} but got no error\n\tTest case: ${_getTestCaseString(
-                    referenceFuncAsync,
-                    values,
-                )}`,
-            );
-        }
-    } catch (e) {
-        if (_.isUndefined(expectedErr)) {
-            throw new Error(`${e.message}\n\tTest case: ${_getTestCaseString(referenceFuncAsync, values)}`);
-        } else {
-            expect(e.message).to.contain(
-                expectedErr,
-                `${e.message}\n\tTest case: ${_getTestCaseString(referenceFuncAsync, values)}`,
-            );
-        }
-    }
-    if (!_.isUndefined(actualResult) && !_.isUndefined(expectedResult)) {
-        expect(actualResult).to.deep.equal(
-            expectedResult,
-            `Test case: ${_getTestCaseString(referenceFuncAsync, values)}`,
-        );
+    // Measure correct behaviour
+    const expected = await evaluatePromise(referenceFuncAsync(...values));
+
+    // Measure actual behaviour
+    const actual = await evaluatePromise(testFuncAsync(...values));
+
+    // Compare behaviour
+    // Note: We could also use `expect(actual).to.deep.equal(expected)` but
+    //       this results in a less readable error message.
+    if (!_.isEqual(actual, expected)) {
+        throw new Error(`Test case ${_getTestCaseString(referenceFuncAsync, values)} expected ${expected} to equal ${actual}`);
```

Comment:
This doesn't work great for objects.

```
> const obj = {a: 'foo', b: 'bar'};
> console.log(`[Value: ${obj}]`);
[Value: [object Object]]
```

If two objects are not equal I wouldn't have any idea how they differ. Can we consider going back to `expect(actual).to.deep.equal(expected)` or maybe using a different technique here?

Reply:
I was try to reuse Chai's `objDisp` function, but it seems that is not publicly accessible. I'll see how expect-equal fares.

Reply:
Ok, now looks like this:

```
      AssertionError: test case {"numerator":"1","denominator":"2","target":"1"}: expected { value: true } to deeply equal { Object (error) }
      + expected - actual

       {
      -  "value": true
      +  "error": "VM Exception while processing transaction: revert UINT256_OVERFLOW"
       }
```